ReactSkia/modules/platform/libcurl/RSkNetworkingModule.cpp

```diff
@@ -89,215 +44,115 @@ uint64_t RSkNetworkingModule::nextUniqueId() {
+    CurlRequest *curlRequestDelegator = (CurlRequest *)userdata;
+    struct NetworkRequest *networkRequest = ((struct NetworkRequest *)userdata);
+    networkRequest->self_->sendProgressEventwrapper(curlRequestDelegator, dltotal, dlnow, ultotal, ulnow);
+    return 0;
+  };
+  auto completionCallback =[&](void* requestObj) -> bool {
+    std::scoped_lock lock(connectionListLock_);
+    CurlRequest *curlRequestDelegator = (CurlRequest *)requestObj;
+    NetworkRequest *networkRequest = (NetworkRequest*)curlRequestDelegator->curldelegator.delegatorData;
+    // completionBlock
+    if(networkRequest) {
+      networkRequest->downloadComplete_ = true;
+      if(!(networkRequest->useIncrementalUpdates_ && (networkRequest->responseType_ == "text") ))
+        sendData(curlRequestDelegator);
+      sendEventWithName("didCompleteNetworkResponse", folly::dynamic::array(curlRequestDelegator->requestId , curlRequestDelegator->curlResponse.errorResult , curlRequestDelegator->curlResponse.responseTimeout ));
+      connectionList_.erase(curlRequestDelegator->requestId);
+      delete networkRequest;
+      delete curlRequestDelegator;
+      return true; //network object is deleted, so curl_cleanup is not required
+    }
+    return false;
+  };
+  curlRequestDelegator_->curldelegator.CURLNetworkingWriteCallback = writeCallback;
+  curlRequestDelegator_->curldelegator.CURLNetworkingProgressCallback = progressCallback;
+  curlRequestDelegator_->curldelegator.CURLNetworkingHeaderCallback = headerCallback;
+  curlRequestDelegator_->curldelegator.CURLNetworkingCompletionCallback = completionCallback;
+  curlRequestDelegator_->curldelegator.delegatorData = networkRequest;
+  curlRequestDelegator_->handle = handle;
+  sharedCurlNetworking_->startCurl(curlRequestDelegator_,query );
+
+  connectionList_[ curlRequestDelegator_->requestId] = curlRequestDelegator_;
   if(callbackObj.isFunction(rt)) {
     jsi::Function callback = callbackObj.getFunction(rt);
-    callback.call(rt, (int) networkRequest->requestId_, 1);
+    callback.call(rt, (int) curlRequestDelegator_->requestId, 1);
   }
-  curl_multi_add_handle(curlMultiHandle_, curl);
-  sem_post(&networkRequestSem_);
   status = jsi::Value((int)CURL_RETURN_SUCESS);
-safe_return :
   if(status.getNumber() == CURL_RETURN_FAILURE && networkRequest )
     delete networkRequest;
   return status;
 }
 
 jsi::Value RSkNetworkingModule::abortRequest(folly::dynamic requestId) {
   std::scoped_lock lock(connectionListLock_);
-  struct NetworkRequest *networkRequest = connectionList_[requestId.asInt()];
-  if(networkRequest == NULL ) {
+  CurlRequest * curlRequest= connectionList_[requestId.asInt()];
+  if(curlRequest == NULL ) {
     RNS_LOG_ERROR ("networkRequest is not valid \n");
     return jsi::Value((int)CURL_RETURN_FAILURE);
   }
-  curl_easy_cleanup(networkRequest->curl_);
-  networkRequest->curl_  = NULL;
-  connectionList_.erase(requestId.asInt());
+  if(sharedCurlNetworking_->abortCurl(curlRequest)) {
```

Comment:
1. move unique id() to network request object.
2. keep curl request object in network request.
3. connection list should maintain network request.
4. change abortcurl to cancelcurl